quickwit/quickwit-indexing/src/source/gcp_pubsub_source.rs

```diff
@@ -0,0 +1,355 @@
+        &mut self,
+        doc_processor_mailbox: &Mailbox<DocProcessor>,
+        ctx: &SourceContext,
+    ) -> Result<Duration, ActorExitStatus> {
+        info!("GcpPubSub beginning batch");
+        let now = Instant::now();
+        let batch_lock = Arc::new(RwLock::new(BatchBuilder::default()));
+        let deadline = time::sleep(*quickwit_actors::HEARTBEAT / 2);
+        tokio::pin!(deadline);
+
+        info!("GcpPubSub pulling batch");
+
+        loop {
+            let mut handles = vec![];
+
+            for _ in 1..self.pull_parallelism {
+                handles.push(self.pull_message_batch(Arc::clone(&batch_lock)));
+            }
+
+            tokio::select! {
+                _ = futures::future::join_all(handles) => {
+                    let batch = batch_lock.read().await;
+                    if batch.num_bytes >= BATCH_NUM_BYTES_LIMIT {
+                        break;
+                    }
+                }
+                _ = &mut deadline => {
+                    break;
+                }
+            }
+            ctx.record_progress();
+        }
+
+        let batch = batch_lock.read().await.clone(); // TODO: This clone is wasteful! There must be a better way
+        if self.should_exit() {
+            info!(subscription = %self.subscription, "Reached end of subscription.");
+            ctx.ask(doc_processor_mailbox, batch.build_force())
+                .await
+                .context("Failed to force commit last batch!")?;
+            self.ack_ids().await?;
+            ctx.send_exit_with_success(doc_processor_mailbox).await?;
+            return Err(ActorExitStatus::Success);
+        }
+
+        if batch.checkpoint_delta.is_empty() {
+            self.state.subscription_is_empty = true
+        } else {
+            info!(
+                num_docs=%batch.docs.len(),
+                num_bytes=%batch.num_bytes,
+                num_millis=%now.elapsed().as_millis(),
+                "Sending doc batch to indexer.");
+            let message = batch.build();
+            ctx.send_message(doc_processor_mailbox, message).await?;
+        }
+
+        Ok(Duration::default())
+    }
+
+    async fn suggest_truncate(
+        &mut self,
+        _checkpoint: SourceCheckpoint,
+        _ctx: &ActorContext<SourceActor>,
+    ) -> anyhow::Result<()> {
+        // TODO: How can we know if these are ok to ack? We need to use the checkpoint!
+        self.ack_ids().await
+    }
+
+    async fn finalize(
+        &mut self,
+        _exit_status: &ActorExitStatus,
+        _ctx: &SourceContext,
+    ) -> anyhow::Result<()> {
+        self.ack_ids().await
+    }
+
+    fn name(&self) -> String {
+        format!(
+            "GcpPubSubSource{{source_id={}}}",
+            self.ctx.source_config.source_id
+        )
+    }
+
+    fn observable_state(&self) -> JsonValue {
+        JsonValue::Object(Default::default())
+    }
+}
+
+impl GcpPubSubSource {
+    async fn ack_ids(&mut self) -> anyhow::Result<()> {
```

Comment:
This also acknowledges batches that are in-flight and have not been published yet. We need to encode our position in the checkpoint that is sent downstream so that when we get a checkpoint back in the suggest truncate, we can ack the published messages only.

I think we can attach a unique ID to each batch sent downstream associated with a set of message IDs to ack in the future:

```rust
type BatchId = Ulid; // Ulids are monotonically increasing
type MessageIDs = Vec<String>

let in_flight_batches: VecDeque<(BatchId, MessageIDs)>

async fn emit_batches(
        &mut self,
        doc_processor_mailbox: &Mailbox<DocProcessor>,
        ctx: &SourceContext,
    ) -> Result<Duration, ActorExitStatus> {
    let now = Instant::now();
    let bach_id = Ulid::new();
    let mut batch_builder = ...;
    let mut message_ids = Vec::new();
    let deadline = time::sleep(*quickwit_actors::HEARTBEAT / 2);
    tokio::pin!(deadline);
    
    loop {
            tokio::select! {
                batch = self.batch_tx.recv() => {
                    for message in batch {
                      batch_builder.add_doc(message.doc);
                      message_ids.push(message.id);
                    }
                    if batch_builder.num_bytes >= BATCH_NUM_BYTES_LIMIT {
                        break;
                    }
                }
                _ = &mut deadline => {
                    break;
                }
            }
            ctx.record_progress();
        }
        let message = batch_builder.build();
        self.in_flight_batches.push((batch_id, message_ids));
         ctx.send_message(doc_processor_mailbox, message).await?;
        }
        Ok(Duration::default())
    }
    
    async fn suggest_truncate(
        &mut self,
        source_checkpoint: SourceCheckpoint
        ctx: &SourceContext,
    ) -> Result<Duration, ActorExitStatus> {
      let max_batch_id = // extract last batch ID published from checkpoint
     // pop batches from `in_flight_batches` from the left as long as batch_id <= max_batch_id and ack them
    }
```

For simplicity, we can do 0 parallelism and no suggest truncate in this PR and implement that in separate PRs.